crates/bevy_pbr/src/render/skin.rs

```diff
@@ -104,20 +196,54 @@ pub fn skins_use_uniform_buffers(render_device: &RenderDevice) -> bool {
         .get_or_init(|| render_device.limits().max_storage_buffers_per_shader_stage == 0)
 }
 
+/// Uploads the buffers containing the joints to the GPU.
 pub fn prepare_skins(
     render_device: Res<RenderDevice>,
     render_queue: Res<RenderQueue>,
-    mut uniform: ResMut<SkinUniforms>,
+    uniform: ResMut<SkinUniforms>,
 ) {
-    if uniform.current_buffer.is_empty() {
+    let uniform = uniform.into_inner();
+
+    if uniform.current_staging_buffer.is_empty() {
         return;
     }
 
-    let len = uniform.current_buffer.len();
-    uniform.current_buffer.reserve(len, &render_device);
-    uniform
-        .current_buffer
-        .write_buffer(&render_device, &render_queue);
+    // Swap current and previous buffers.
+    mem::swap(&mut uniform.current_buffer, &mut uniform.prev_buffer);
+
+    // Resize the buffer if necessary. Include extra space equal to `MAX_JOINTS`
+    // because we need to be able to bind a full uniform buffer's worth of data
+    // if skins use uniform buffers on this platform.
+    let needed_size = (uniform.current_staging_buffer.len() as u64 + MAX_JOINTS as u64)
+        * size_of::<Mat4>() as u64;
```

Comment:
Couldn't it just be `.max(MAX_JOINTS)`?

Reply:
The issue is that, in CPU-driven mode, the uniform buffer is a sliding window that has to be exactly `MAX_JOINTS` in size. So if the start of the window is placed at the end, then we have to have `MAX_JOINTS - 1` extra entities at the end. I guess this can actually be `uniform.current_starting_buffer.len().max(1) as u64 + MAX_JOINTS as u64 - 1)`.

Reply:
This is fine enough.